### PR TITLE
Add duplicate logs check.

### PR DESCRIPTION
**Pull request by jefchien**
# Description of the issue
Need to be able to detect duplicate logs in the integration tests.

# Description of changes
Adds a duplicate log check to the basic validator. Checks that there aren't any logs for the same timestamp and message. Refactored the log validation a bit to evaluate over LogEvents instead of the log messages to retain the timestamp.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Ran the integration tests and failed on the duplicates.

### CODE REVIEW

File: test/ecs/ecs_metadata/ecs_metadata_test.go [hunk 1]
```diff
@@ -57,25 +55,25 @@ func TestValidatingCloudWatchLogs(t *testing.T) {
 
 		end := time.Now()
 
-		ok, err := awsservice.ValidateLogs(logGroupName, LogStreamName, &start, &end, func(logs []string) bool {
```
Comment (adam-mateen) [inline test/ecs/ecs_metadata/ecs_metadata_test.go:60]: Having an integration test for ECS metadata which relies on the agent's own log getting uploaded to CWL is not very robust. Log messages can change at a developer's whim...
But if there are no better options then okay.

File: util/awsservice/cloudwatchlogs.go [hunk 2]
```diff
@@ -161,15 +168,97 @@ func GetLogStreams(logGroupName string) []types.LogStream {
...
+			if !ok {
+				messages = map[string]struct{}{}
+				byTimestamp[timestamp] = messages
+			}
+			_, ok = messages[message]
+			if !ok {
+				messages[message] = struct{}{}
+				continue
+			}
+			return fmt.Errorf("duplicate message found at %v | message: %s", time.UnixMilli(timestamp), message)
```
Comment (adam-mateen) [inline util/awsservice/cloudwatchlogs.go:260]: nit: we could avoid the continue by doing:
```
_, ok = messages[message]
if ok {
	return fmt.Errorf("duplicate message found at %v | message: %s", time.UnixMilli(timestamp), message)
}
messages[message] = struct{}{}
```

File: util/awsservice/cloudwatchlogs.go [hunk 3]
```diff
@@ -161,15 +168,97 @@ func GetLogStreams(logGroupName string) []types.LogStream {
...
+		if len(events) != count {
+			return fmt.Errorf("actual log events count (%v) does not match expected (%v)", len(events), count)
+		}
+		return nil
+	}
+}
+
+func AssertNoDuplicateLogs() LogEventsValidator {
+	return func(events []types.OutputLogEvent) error {
+		byTimestamp := make(map[int64]map[string]struct{})
```
Comment (adam-mateen) [inline util/awsservice/cloudwatchlogs.go:246]: I guess there could be an agent bug reporting the same event multiple times with different timestamps... Like rereading a file or rereading an event log. But it is hard (or impossible) to distinguish this scenario from an event that legitimately repeats at different times.

File: util/awsservice/cloudwatchlogs.go [hunk 4]
```diff
@@ -161,15 +168,97 @@ func GetLogStreams(logGroupName string) []types.LogStream {
...
+		if err != nil {
+			return fmt.Errorf("failed to execute schema validator: %w", err)
+		} else if len(keyErrors) > 0 {
+			return fmt.Errorf("failed schema validation: %v | schema: %s | log: %s", keyErrors, schema, message)
+		}
+		return nil
+	}
+}
+
+func AssertLogSubstring(substr string) LogEventValidator {
```
Comment (adam-mateen) [inline util/awsservice/cloudwatchlogs.go:203]: nit: The name did not make it clear if this was asserting the log contains the substring or does not.
We could call it `AssertLogContainsSubstring`

File: util/awsservice/cloudwatchlogs.go (hunk 3 above)
Comment (jefchien) [reply util/awsservice/cloudwatchlogs.go:246]: Yeah we repeat the same logs quite often for tests, so it'd be harder to catch that.

**Review by adam-mateen — APPROVED** [review]

**Review by sky333999 — APPROVED** [review]

**Review by okankoAMZ — APPROVED** [review]

**Pull request merged**